0013-RS-pubsub-subscription-streaming.md

```diff
@@ -0,0 +1,142 @@
+  //
+  // metadata property:
+  // - key : the key of the message.
+  map<string, string> metadata = 3;
+
+  // dead_letter_topic is the topic to which messages that fail to be processed
+  // are sent.
+  optional string dead_letter_topic = 4;
+}
+
+// SubscribeTopicEventsResponse is a message containing the result of a
+// subscription to a topic.
+message SubscribeTopicEventsResponse {
+  // id is the unique identifier for the subscription request.
+  string id = 1;
+
+  // status is the result of the subscription request.
+  TopicEventResponse status = 2;
+}
+```
+
+When an application wishes to subscribe to a topic it will initiate a stream with `SubscribeTopicEventsRequest`, and `Send` the initial request `SubscribeTopicEventsInitialRequest` containing the options for the subscription.
+Daprd will then setup the machinery to add this gRPC RPC stream to the set of subscribers.
+The request contains no route or path matching configuration as all events will be sent on this stream.
+Subscription gRPC streams are the highest priority when Daprd determines which publisher a message should be sent to.
+Only a single PubSub Topic pair may be subscribed at a single time with this API.
+If the first message sent to the server is not the initial request, the RPC will return an error.
+If any subsequent messages are not `SubscribeTopicEventsResponse` messages, the RPC will return an error.
+
+When a message is published to the topic, Daprd will send a `TopicEventRequest` message on the stream containing the message payload and metadata.
+After the application has processed the message, it will send to the server a `SubscribeTopicEventsResponse` containing the `id` of the message and the `status` of the message processing.
+Since multiple messages can be sent and processed in the application at the same time, the event `id` is used by the server to track the status of each individual event.
+An event topic response will follow the timeout resiliency as currently exist for subscriptions.
+
+Client code:
+
+```go
+	stream, _ := client.SubscribeTopicEvents(ctx)
+	stream.Send(&rtv1.SubscribeTopicEventsRequest{
+		SubscribeTopicEventsRequestType: &rtv1.SubscribeTopicEventsRequest_InitialRequest{
+			InitialRequest: &rtv1.SubscribeTopicEventsInitialRequest{
+				PubsubName: "mypub", Topic: "a",
+			},
+		},
+	})
+
+	client.PublishEvent(ctx, &rtv1.PublishEventRequest{
+		PubsubName: "mypub", Topic: "a",
+		Data:            []byte(`{"status": "completed"}`),
+		DataContentType: "application/json",
+	})
+
+	event, _ := stream.Recv()
+	stream.Send(&rtv1.SubscribeTopicEventsRequest{
+		SubscribeTopicEventsRequestType: &rtv1.SubscribeTopicEventsRequest_EventResponse{
+			EventResponse: &rtv1.SubscribeTopicEventsResponse{
+				Id:     event.Id,
+				Status: &rtv1.TopicEventResponse{Status: rtv1.TopicEventResponse_SUCCESS},
+			},
+		},
+	})
+
+	stream.CloseSend()
+```
+
+## Completion Checklist
+
+- [ ] gRPC server implementation in daprd
```

Comment:
I think we need a HTTP implementation too. Would probably be 2 APIs:

1. A method to receive events, which uses server-sent events
2. A method to confirm the completion of a message

Reply:
I am OK for this to be gRPC only since we have another way to subscribe for HTTP already.

Reply:
That's how we started with Configuration API I reckon, being gRPC only, but I would like to see a separate proposal for HTTP too, independent of this one.

Reply:
When working on the crypto building block I was told it had to support HTTP too in the first iteration.

We also have another way to subscribe for gRPC too... But this way is arguably better (and would be better for HTTP too since it doesn't require opening ports in the app).

Reply:
There's a distinction here in that this isn't a new building block but instead additional methods into an already existing building block where users have options to subscribe with HTTP already

Reply:
https://github.com/dapr/proposals/blob/main/guides/api-design.md#updates-to-existing-apis

<img width="907" alt="image" src="https://github.com/dapr/proposals/assets/43508/11b64f6e-2709-4da5-9e51-f70f16f1efd9">

Reply:
This is exactly what I had in mind when I asked for a proposal for the HTTP APIs also, but these can be separate and independent still

Reply:
I am in favor of a proposal for HTTP that does not use streaming, solving the problem of dynamic subscription for HTTP but still use the HTTP callback like we did for configuration API.

Reply:
I've updated the proposal with an HTTP equivalent, PTAL!

Reply:
cc @mukundansundar as per our discussion